x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/In.java

```diff
@@ -139,4 +168,184 @@ protected TypeResolution resolveType() { // TODO: move the foldability check fro
+        }
+        if (commonType == INTEGER) {
+            return new InIntEvaluator.Factory(source(), lhs, factories);
+        }
+        if (commonType == LONG || commonType == DATETIME || commonType == UNSIGNED_LONG) {
+            return new InLongEvaluator.Factory(source(), lhs, factories);
+        }
+        if (commonType == KEYWORD
+            || commonType == TEXT
+            || commonType == IP
+            || commonType == VERSION
+            || commonType == UNSUPPORTED
+            || EsqlDataTypes.isSpatial(commonType)) {
+            return new InBytesRefEvaluator.Factory(source(), toEvaluator.apply(value), factories);
+        }
+        if (commonType == NULL) {
+            return EvalOperator.CONSTANT_NULL_FACTORY;
+        }
+        throw EsqlIllegalArgumentException.illegalDataType(commonType);
+    }
+
+    private DataType commonType() {
+        DataType commonType = value.dataType();
+        for (Expression e : list) {
+            if (e.dataType() == NULL && value.dataType() != NULL) {
+                continue;
+            }
+            if (EsqlDataTypes.isSpatial(commonType)) {
+                if (e.dataType() == commonType) {
+                    continue;
+                } else {
+                    commonType = NULL;
+                    break;
+                }
+            }
+            commonType = EsqlDataTypeRegistry.INSTANCE.commonType(commonType, e.dataType());
+        }
+        return commonType;
+    }
+
+    static void process(BooleanBlock.Builder builder, BitSet nulls, BitSet mvs, boolean lhs, boolean[] rhs) {
+        boolean hasNull = nulls == null ? false : nulls.cardinality() > 0;
+        for (int i = 0; i < rhs.length; i++) {
+            if (hasNull && nulls.get(i)) {
+                continue;
+            }
+            if (mvs != null && mvs.get(i)) {
+                continue;
+            }
+            Boolean compResult = Comparisons.eq(lhs, rhs[i]);
+            if (compResult == Boolean.TRUE) {
+                builder.appendBoolean(true);
+                return;
+            }
+        }
+        if (hasNull) {
+            builder.appendNull();
+        } else {
+            builder.appendBoolean(false);
+        }
+    }
+
+    static void process(BooleanBlock.Builder builder, BitSet nulls, BitSet mvs, int lhs, int[] rhs) {
+        boolean hasNull = nulls == null ? false : nulls.cardinality() > 0;
+        for (int i = 0; i < rhs.length; i++) {
+            if (hasNull && nulls.get(i)) {
+                continue;
+            }
+            if (mvs != null && mvs.get(i)) {
+                continue;
+            }
+            Boolean compResult = Comparisons.eq(lhs, rhs[i]);
+            if (compResult == Boolean.TRUE) {
+                builder.appendBoolean(true);
+                return;
+            }
+        }
+        if (hasNull) {
+            builder.appendNull();
+        } else {
+            builder.appendBoolean(false);
+        }
+    }
```

Comment:
Are these worth moving to the generated code?

Reply:
> Are these worth moving to the generated code?

That's a good idea! I'll move the builder append call into the generated code.